flink-runtime/src/main/java/org/apache/flink/runtime/leaderelection/LeaderElectionEventHandler.java

```diff
@@ -36,14 +37,44 @@ public interface LeaderElectionEventHandler {
      * Called by specific {@link LeaderElectionDriver} when the leadership is granted.
      *
      * @param newLeaderSessionId the valid leader session id
+     * @param executorService the {@link ExecutorService} this call should be executed on.
      */
-    void onGrantLeadership(UUID newLeaderSessionId);
+    void onGrantLeadershipAsync(UUID newLeaderSessionId, ExecutorService executorService);
+
+    /**
+     * Called by specific {@link LeaderElectionDriver} when the leadership is granted.
+     *
+     * @param newLeaderSessionId the valid leader session id
+     */
+    void onGrantLeadershipAsync(UUID newLeaderSessionId);
```

Comment:
I'm a bit confused, why is this method named with async suffix? I know this is to make `DefaultMultipleComponentLeaderElectionService` use the executor in `DefaultLeaderElectionService`, but from the protocol of the interface, it's a bit difficult to understand. At least we can emphasize in Java doc that this method must not be executed on the caller thread.

The following is the doc for the `thenRunAsync` method of `CompletionStage`, which looks much more intuitive.
```
    /*
     * Returns a new CompletionStage that, when this stage completes
     * normally, executes the given action using this stage's default
     * asynchronous execution facility.
     */
    public CompletionStage<Void> thenRunAsync(Runnable action);
```

Reply:
You're right - iterating over this change once more makes me realize that it's a bad proposal. The directExecutor doesn't need to be specified through a parameter. Instead, we could keep `onGrantLeadership(UUID)` (analogously for the other methods) to call the logic synchronously. This can be used by the legacy `LeaderElectionDriver` implementations.

The asynchronous call that's going to be triggered from the `DefaultMultipleComponentLeaderElectionService` could be named something like `forwardGrantLeadershipEvent(UUID)` (analogously for the other methods). In its JavaDoc we have to define that this method is expected to run in a separate thread.

I'm gonna go ahead a work on a commit adjusting the interface tomorrow.